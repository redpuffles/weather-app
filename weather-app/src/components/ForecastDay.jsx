/**
 * A component that displays brief forecast weather data for a day.
 * 
 * Shows the day name, temperature, and a weather conditions icon.
 * 
 * @file This file defines the ForecastDay component class.
 * @author Roger.
 * @since 1.0.3
 */

import PropTypes from "prop-types";
import React from "react";

import * as S from "../styles/ForecastDay.styles";

function ForecastDay (props) {
  const [alt, setAlt] = React.useState(""); // Alt text for the icon image.
  const [day, setDay] = React.useState(""); // The day this component corresponds to, shortened to three letters.
  const [iconURL, setIconURL] = React.useState(""); // // An icon URL reflecting forecasted weather conditions.
  const [temp, setTemp] = React.useState(""); // Forecasted average temperature.

  /**
   * Stores and updates alt from props.
   */
  React.useEffect(() => {
    if (props.alt !== "") {
      setAlt(props.alt);
    } else {
      setAlt("Alt text unavailable.");
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.alt]);

  /**
   * Stores and updates day from props (a day should always be provided).
   */
  React.useEffect(() => {
    setDay(props.day);
  }, [props.day]);

  /**
   * Stores and updates icon URL from props. Convert props.icon into the
   * correspnding URL.
   */
  React.useEffect(() => {
    if (props.icon !== "") {
      setIconURL(props.getIconURL(props.icon));
    } else {
      setIconURL("");
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.icon]);

  /**
   * Stores and updates temp from props.
   */
  React.useEffect(() => {
    if (props.Temp !== "") {
      setTemp(Math.round(props.temp));
    } else {
      setTemp("-");
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [props.temp]);

  return (
    <S.ForecastDay aria-label={day} role="forecast-day">
      <S.Day data-testid="day">{day}</S.Day>
      <S.Temp data-testid="temp">{temp}Â°</S.Temp>
      {iconURL !== ""
        ? <S.Icon alt={alt} data-testid="icon" src={iconURL}/>
        : null
      }
    </S.ForecastDay>
  );
}

ForecastDay.propTypes = {
  alt: PropTypes.string.isRequired,
  day: PropTypes.string.isRequired,
  getIconURL: PropTypes.func.isRequired,
  icon: PropTypes.string.isRequired,
  temp: PropTypes.string.isRequired
}

export default ForecastDay;
