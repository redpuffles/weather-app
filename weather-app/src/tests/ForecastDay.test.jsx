/**
 * React unit tests for ForecastDay.
 * 
 * - "renders properly given the minimal required props"
 * - "renders properly given relevant weatherData props"
 * 
 * @file This file contains unit tests for the ForecastDay component.
 * @author Roger.
 * @since 1.0.0
 */

import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import renderer from 'react-test-renderer';
import { screen } from '@testing-library/react';

import ForecastDay from "../components/ForecastDay";

let container = null;

afterEach(() => {
  // Clean up after exit.
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

beforeEach(() => {
  // Set up a DOM element as a render target.
  container = document.createElement("div");
  document.body.appendChild(container);
});

it("renders properly given the minimal required props", () => {
  act(() => {
    render(<ForecastDay
              alt={""}
              day={"test-day"}
              getIconURL={() => {}}
              icon={""}
              temp={""}
           />, container);
  });
  expect(screen.getByTestId("day")).toHaveTextContent("test-day");
  expect(screen.getByTestId("temp")).toHaveTextContent("°");
  
  // snapshot rendering with minimal props
  const test = renderer.create(<ForecastDay
                                  alt={""}
                                  day={"Day"}
                                  getIconURL={() => {}}
                                  icon={""}
                                  temp={""}
                               />, container);
  expect(test).toMatchSnapshot();
});

it("renders properly given relevant weatherData props", () => {
  // given prop values should be reflected, iconURL should be called
  const mockURL = jest.fn();

  act(() => {
    render(<ForecastDay
              alt={"test-alt"}
              day={"test-day"}
              getIconURL={mockURL}
              icon={"01d"}
              temp={"1"}
           />, container);
  });
  expect(screen.getByTestId("day")).toHaveTextContent("test-day");
  expect(screen.getByTestId("temp")).toHaveTextContent("1°");
  expect(screen.getByAltText("test-alt")).toBeTruthy();
  expect(mockURL).toHaveBeenCalledTimes(1);
});
