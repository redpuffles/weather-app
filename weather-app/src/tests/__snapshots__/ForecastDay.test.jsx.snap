// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly given the minimal required props 1`] = `
<div
  aria-label=""
  className="sc-gtsrHT kapWDu"
  role="forecast-day"
>
  <span
    className="sc-bdnxRM iFkTIe"
    data-testid="day"
  >
    
  </span>
  <span
    className="sc-hKFxyN hDxXs"
    data-testid="temp"
  >
    
    Â°
  </span>
</div>
`;
